gtsam/basis/ParameterMatrix.h

```diff
@@ -0,0 +1,215 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file ParamaterMatrix.h
```

Comment:
Rethinking this. Do we really need this? maybe we should just use `Eigen::Matrix<double, M, -1>` and kill this file.

Reply:
I've already explained this in a previous comment. Pasting here verbatim :slightly_smiling_face: 

Doing 
```cpp
template <int M>
using ParameterMatrix = Eigen::Matrix<double, M, -1>
```
causes the API to become far more verbose and non-intuitive.

E.g., we can no longer do `ParameterMatrix<12> X(N)` since it gives an error that we tried calling a vector method on a matrix object. We have to instead do
```cpp
ParameterMatrix<12> X;
X.resize(12, N); // why do we need to specify 12 again here? This is something eigen needs :(
```

In general, the fact that one of the dimensions is dynamic and the other has size greater than 1 makes things non-standard for Eigen since it assumes  everything is a Matrix and simple things like `X.col` don't compile.

I'll add this to the docstring for future reference.

Reply:
Our other option is is to just use `Matrix<double, -1, -1>` but then we lose compile time checks on the matrix size.

Reply:
It should be very easy to make a small function that creates the right matrix. Could we replac entire file contents with code below?

```
template <int M>
using ParameterMatrix = Eigen::Matrix<double, M, -1>;

template <int M>
ParameterMatrix<M> MakeParameters(int  N) {
  ParameterMatrix<M> P;
  P.resize(M, N);
  return P;
}

```

and say 

```
auto P = MakeParameters<12>(100);
```

Reply:
That solves the initialization problem but not the various vector-matrix confusion errors that Eigen throws on compilation.

I did the whole typedef and actually using the matrix, and it didn't compile.